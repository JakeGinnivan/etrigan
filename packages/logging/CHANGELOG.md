# @etrigan/logging

## 0.1.0
### Minor Changes

- 13e0e6e: Initial release
