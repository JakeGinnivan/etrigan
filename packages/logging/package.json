{
  "name": "@etrigan/logging",
  "version": "1.0.0",
  "description": "Etrigan logging",
  "main": "dist/index.js",
  "author": "Seven West Media WA",
  "license": "MIT",
  "peerDependencies": {
    "tslib": "^1.10.0"
  },
  "dependencies": {
    "pino": "^5.13.6",
    "uuid": "^3.3.3"
  }
}
