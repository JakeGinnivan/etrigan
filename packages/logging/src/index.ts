export { WithLoggingInfo, Levels, createLogger } from './create-logger'
export { expressRequestLoggingMiddleware } from './logging-middleware'
export { scrubJsonLogs, scrubPrettyLogs } from './test-helpers/log-scrubber'
