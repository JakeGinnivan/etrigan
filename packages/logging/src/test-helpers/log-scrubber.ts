export function scrubPrettyLogs(msg: string): string
export function scrubPrettyLogs(msg: string[]): string[]
export function scrubPrettyLogs(msg: string | string[]) {
    if (typeof msg === 'string') {
        return msg
            .replace(/\((.*?\/)?\d{3,5} on .*?\)/, '($100000 on <hostname>)')
            .replace(/\w{8}-\w{4}-\w{4}-\w{4}-\w{12}"/g, '00000000-0000-0000-0000-000000000000')
            .replace(/ \d{13}/gm, ' 0000000000000')
            .replace(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/g, '0000-00-00T00:00:00.000Z')
            .replace(/responseTime: \d+/g, 'responseTime: 000')
            .replace(/127.0.0.1:\d+/g, '127.0.0.1:<port>')
    }

    return msg.map(m => scrubPrettyLogs(m))
}
