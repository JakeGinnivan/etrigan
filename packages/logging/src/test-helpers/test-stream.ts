import { Writable } from 'stream'

export class TestStream extends Writable {
    constructor(private onWrite: (msg: string) => void) {
        super()

        const newWrite: Writable['_write'] = (chunk, _encoding, callback) => {
            this.onWrite(chunk.toString())
            callback()
        }
        this._write = newWrite
    }
}
