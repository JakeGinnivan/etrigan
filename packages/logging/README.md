# Etrigan Logging

Opinionated helpers for production logging using express and pino.

Has the following features

-   Express middleware: logs request begin/end messages, log sampling and more.
-   Log escalator: buffers request logs, enabling debug/info level logs to be written only when warn/error messages are also logged ensuring you do not miss valuable context before a error occurs.
-   Pino serialisers: for express requests

See below for more information on these features

## Express request logging middleware

Will automatically log 'begin request' at debug level then close or finish at info level.

### Features

#### Enable debug logs query string

Sometimes it's useful to get debug level logs in production, by default adding `?debug_log` as a query string on the request will automatically set the log level for the request to debug.

#### Log sampling

You can specify what % of logs to log at debug and what level at info. This helps reduce overall log volumes while still giving you debug and info level logs to gather metrics and other useful information from.

The log escalator, explained below will ensure that contextual info or debug level logs are also logged when warns or errors are logged.

#### Log

### Usage

```ts
import { expressRequestLoggingMiddleware } from '@etrigan/logging'

// On your express server
app.use(expressRequestLoggingMiddleware(log))
```

### Options

You can also add `?debug_log` to any API call to increase the log level for that request to debug. Setting the env variable `DISABLE_DEBUG_LOG_QS` will turn this feature off.

## Log escalalator

When sampling logs only warns and above will be logged by default. This is problematic when errors occur because they may not occur on a sampled log.

The log escalator solves this problem. It will ensure when a warn or error is logged that the debug and info level logs which preceeded the error will also be logged.

### How it works

-   It buffers the log messages for a request
-   When a warn is logged, the request logger will be set to `info` level
-   When a error is logged, the request logger will be set to `error` level
-   Once the request finishes, errors or gets closed the buffered logs will get written to the logger
-   It will automatically time out after 10 seconds and flush the buffered logs regardless.
