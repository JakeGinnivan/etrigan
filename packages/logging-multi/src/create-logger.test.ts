import { createLogger } from '.'
import { TestStream } from './test-helpers/test-stream'
import { scrubJsonLogs } from '@etrigan/logging'

it('can create logger', () => {
    const logs: string[] = []
    const log = createLogger({
        name: 'named',
        logLevel: 'debug',
        additionalStreams: [{ stream: new TestStream(msg => logs.push(msg)) }],
    })

    log.info('Test')

    expect(scrubJsonLogs(logs)).toMatchSnapshot()
})
