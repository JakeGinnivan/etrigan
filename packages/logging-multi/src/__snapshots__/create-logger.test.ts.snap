// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create logger 1`] = `
Array [
  "{\\"level\\":30,\\"time\\":someunixtimestamp,\\"pid\\":00000,\\"hostname\\":\\"replaced\\",\\"name\\":\\"named\\",\\"msg\\":\\"Test\\",\\"v\\":1}
",
]
`;
